presto-main/src/main/java/io/prestosql/operator/scalar/ArrayCombinationsFunction.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.block.PageBuilderStatus;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.function.TypeParameter;
+import io.prestosql.spi.type.ArrayType;
+import io.prestosql.spi.type.Type;
+
+import static io.prestosql.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.prestosql.spi.type.StandardTypes.INTEGER;
+import static io.prestosql.util.Failures.checkCondition;
+import static java.lang.StrictMath.toIntExact;
+import static java.util.Arrays.setAll;
+
+@ScalarFunction("combinations")
+@Description("Return k-element subsets from array")
+public final class ArrayCombinationsFunction
+{
+    private ArrayCombinationsFunction() {}
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block combinations(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block array,
+            @SqlType(INTEGER) long k)
+    {
+        int n = array.getPositionCount();
+        checkCondition(k >= 0, INVALID_FUNCTION_ARGUMENT, "k must be positive");
+
+        ArrayType arrayType = new ArrayType(elementType);
+        if (k > n) {
+            return arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 0).build();
+        }
+
+        BlockBuilder resultBuilder = arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 10_000 /* TODO n! / k! (n-k)! , but mind overflow */);
+        int[] combination = firstCombination(toIntExact(k));
+        do {
+            BlockBuilder arrayBuilder = elementType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), toIntExact(k));
+            for (int position : combination) {
+                elementType.appendTo(array, position, arrayBuilder);
+            }
+            arrayType.writeObject(resultBuilder, arrayBuilder.build());
```

Comment:
Instead of creating an ArrayBlock for each combination and appending to the outer array, you can use beginEntry/closeEntry to write them once. Additionally, we could use a dictionary block for the values so that the values in the original array appear just once.